kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,442 @@
+            let actions = actions?;
+
+            // Note: We do not use a [`DataSkippingFilter`] because we need to visit all add and
+            // remove actions for deletion vector resolution to be correct.
+            //
+            // Consider a scenario with a pair of add/remove actions with the same path. The add
+            // action has file statistics, while the remove action does not (stats is optional for
+            // remove). The DataSkippingFilter may skip the add action, and the remove action will
+            // remain. The data file will be read for the remove action, which is unnecessary because
+            // all of the rows will be filtered by the predicate. Instead, we wait until deletion
+            // vectors are resolved so that we can skip both actions in the pair.
+            let mut visitor = PreparePhaseVisitor {
+                add_paths: &mut add_paths,
+                remove_dvs: &mut remove_dvs,
+                has_cdc_action: &mut has_cdc_action,
+                protocol: None,
+                metadata_info: None,
+            };
+            visitor.visit_rows_of(actions.as_ref())?;
+
+            if let Some(protocol) = visitor.protocol {
+                ensure_cdf_read_supported(&protocol)
+                    .map_err(|_| Error::change_data_feed_unsupported(commit_file.version))?;
+            }
+            if let Some((schema, configuration)) = visitor.metadata_info {
+                let schema: StructType = serde_json::from_str(&schema)?;
+                // Currently, schema compatibility is defined as having equal schema types. In the
+                // future, more permisive schema evolution will be supported.
+                // See: https://github.com/delta-io/delta-kernel-rs/issues/523
+                require!(
+                    table_schema.as_ref() == &schema,
+                    Error::change_data_feed_incompatible_schema(table_schema, &schema)
+                );
+                let table_properties = TableProperties::from(configuration);
+                check_cdf_table_properties(&table_properties)
+                    .map_err(|_| Error::change_data_feed_unsupported(commit_file.version))?;
+            }
+        }
+        // We resolve the remove deletion vector map after visiting the entire commit.
+        if has_cdc_action {
+            remove_dvs.clear();
+        } else {
+            // The only (path, deletion_vector) pairs we must track are ones whose path is the
+            // same as an `add` action.
+            remove_dvs.retain(|rm_path, _| add_paths.contains(rm_path));
+        }
+        Ok(LogReplayScanner {
+            timestamp: commit_file.location.last_modified,
+            commit_file,
+            has_cdc_action,
+            remove_dvs,
+        })
+    }
+    /// Generates an iterator of [`TableChangesScanData`] that iterates over each action of the
+    /// commit. This performs phase 2 of [`LogReplayScanner`].
+    fn into_scan_batches(
```

Comment:
i still have a little bit of hesitancy to the design of `LogReplayScanner` doing a 'prepare phase' constructor, then immediately doing the `into_scan_batches`. We aren't holding on to any useful state and the only call site is just doing one followed by the next right? I have been thinking about a better solution but perhaps if the current one works we take it as a cleanup follow-up? Even if we just make a "think about a better way to do table_changes LogReplayScanner" issue

Reply:
I'm thinking of maybe just a simple function that does:
1. read the actions from JSON
2. do a first iteration over them for the prepare phase
3. do a second iteration (map) over them for the 'yield' phase

then the double-iteration is clear, and you have both pieces next to each other? maybe just an idea to think about and optimize later :)

Reply:
Fair, Perhaps this didn't need to be a struct. Happy to take this up as a clean-up followup. I do worry about having it as a single function because that can get pretty long. Let's discuss at some point 👍

Reply:
After a bit of pondering, here's my take: The current code is a good internal structure (software engineering, rule of 30, etc). It's not a great user surface (see e.g. unit tests all having to do the same boilerplate, see also the worries about exposing `DataSkippingPredicate` to FFI). 

What we're missing is a top-level function that encapsulates all this and presents a good user surface. Possibly as simple as defining a `table_changes::table_changes_action_iter` to match our existing `scan::scan_action_iter`?
(note that `scan::LogReplayScanner` is not even `pub(crate)`, let alone `pub`).

Reply:
@scovich Do you think it would be better to have the tests interact with `table_changes_action_iter`? I went one level down because I only need to test one commit at a time.

Reply:
Going by "test what you run in prod" -- yes we should use the method. It would certainly not be a rare case to process a single commit in real life.